Scripted tests, bugfixes, and default bump.

jrouly opened the pull request:
This PR does two things:
* bumps `sbt-release` to the latest version
* introduces a simple scripted test for the plugin

Note: the test is _failing_ locally right now. Either I'm misunderstanding the correct use of this plugin (in which case I will be happy to update the test scenario) or there's a bug that needs tracking down (which I'm also happy to assist with). Let me know what your take on the situation is. Either way, I think scripted tests are valuable for this plugin.

The scenario itself is pretty basic. It starts at `0.0.0` with `version.sbt` committed and a tag `v0.0.0` present. Then it makes a `fix:` commit, a `minor:` commit, and a `major:` commit. And it expects the version to bump to `0.0.1-SNAPSHOT`, `0.1.0-SNAPSHOT`, and `1.0.0-SNAPSHOT` respectively.

Right now the failure is that, after a `fix:` commit, the version is bumped to `0.1.1-SNAPSHOT`....which is not what I would expect.

FYI @pdalpra @sullis

jrouly commented on src/main/scala/autoversion/AutoVersionPlugin.scala:
```diff
@@ -30,7 +30,7 @@ object AutoVersionPlugin extends AutoPlugin {
     suggestedBump := suggestBump.value,
     releaseVersion := { ver =>
-      Version(ver).map(v => v.bump(suggestedBump.value).withoutQualifier.string).getOrElse(versionFormatError)
+      Version(ver).map(v => v.bump(suggestedBump.value).withoutQualifier.string).getOrElse(versionFormatError(ver))
```
AFAICT this is the only change in `sbt-release` between 1.0.6 and 1.0.15 that needs addressing.

The test fails on 1.0.6 as well, so this isn't critical for the PR.

pdalpra commented:
Hi, thanks for your PR !
Your tests seem just fine, I think I found what causes the failures: [The default regexes for minor bumps](https://github.com/sbt/sbt-autoversion/blob/master/src/main/scala/autoversion/AutoVersionPlugin.scala#L27) is much too lenient, and therefore match any commit.
This behavior of matching any other commit message as a minor bump was inspired from [cookbook-release(https://github.com/criteo/cookbook-release), but this means that we should try to match minor bumps last, as they are the catch-all case.
Therefore, tweaking [Commit.suggestedBump](https://github.com/sbt/sbt-autoversion/blob/master/src/main/scala/autoversion/model/Commit.scala#L14-L26) to account for that should make your tests pass :)

jrouly commented:
@pdalpra Yep, that sounds right. I'll look into it.

It might be nice to add a `SettingKey[List[Version.Bump]]` to allow users to change the order in case they want a different catch-all, but that may also totally be overkill.

jrouly commented on .travis.yml:
```diff
@@ -3,7 +3,7 @@ jdk:
 
 script:
- - sbt ^test ^publishLocal
+ - sbt ^test ^scripted ^publishLocal
```
Updated CI to run scripted tests.

jrouly commented on README.md:
```diff
@@ -41,6 +41,8 @@ Default patterns:
 * minor: `.*`
 * bugfix: `\[?bugfix\]?.*`, `\[?fix\]?.*`
 
+The default minor regex will match any commits which are not otherwise specified.
```
Added an explicit note in the README that `minor` acts as a catch-all.

jrouly commented on src/sbt-test/sbt-autoversion/simple/test:
```diff
@@ -0,0 +1,41 @@
+# Anticipate auto version.
+> reload
+> release with-defaults
+$ exec grep -Fq "0.0.2-SNAPSHOT" version.sbt
```
Switched to using `grep` instead of `must-mirror` because `0.13`/`1.x` use different syntax for scoping the version. It also makes the scenario easier to read in-place :shrug:

jrouly commented on src/sbt-test/sbt-autoversion/simple/test:
```diff
@@ -0,0 +1,41 @@
+# Anticipate auto version.
+> reload
+> release with-defaults
+$ exec grep -Fq '"0.0.2-SNAPSHOT"' version.sbt
```
Switched to using `grep` instead of `must-mirror` because `0.13`/`1.x` use different syntax for scoping the version. It also makes the scenario easier to read in-place :shrug:

jrouly commented:
Identified another bug. Any time `bugfix` and `minor` both match, `suggestedBumps.max` will return `minor` -- but `minor` _always_ matches.

jrouly commented:
There's another issue I'm running into where patch versions are getting skipped. For example:

```
0.0.0
fix: foobar
0.0.1
0.0.2-snapshot
fix: foobar
0.0.3
0.0.4-snapshot
```

A bugfix on a patch snapshot jumps the next patch version (`0.0.2` is never released).

pdalpra commented:
Thanks for all your debugging 👍 
I’ll be honest, I’m not sure that plugin ever saw real production usage yet, so I’m not that surprised that you found some bugs :/

jrouly commented:
@pdalpra haha, no problem! I was just happy to discover someone had already started work on a plugin like this. I'm coming from [semantic-release](https://www.npmjs.com/package/semantic-release), and this feels very similar.

I've pushed some commits to address the bug I described, by the way. It was the only edge case I could identify, so I went ahead and just added an explicit case for it (and unit tests as well).

I also switched the catch-all regex from `minor` to `patch`. `minor` now looks for `feature|minor`. This lines up with `semantic-release` and other javascript "auto version" approaches. In general I think it makes more sense for the "default" (catch-all) behavior to be `patch`, anyway.

If I can get your blessing, I would also like to adjust the regexes and readme slightly to line up with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/). That would probably be a separate PR though. Let me know what you think!

jrouly commented on src/main/scala/autoversion/AutoVersion.scala:
```diff
@@ -0,0 +1,19 @@
+    // Otherwise the expected bugfix version will be skipped.
+    // e.g. 0.1.4 => 0.1.5-SNAPSHOT => 0.1.5
+    Version(ver).map {
+      case v if v.qualifier.isDefined && bugfixSuggested => v.withoutQualifier.string
```
This is the fix for skipped patch versions.

The next `case` is the old behavior, and it applies in every other situation.

jrouly commented on src/sbt-test/sbt-autoversion/simple/test:
```diff
@@ -0,0 +1,104 @@
+$ exec grep -Fq '"2.0.1-SNAPSHOT"' version.sbt
+
+# 2.0.1-SNAPSHOT => 2.0.1 => 2.0.2-SNAPSHOT
+$ exec git commit --allow-empty -m 'doing hard work'
```
I also added an example of an unconventional commit, i.e. one that exercises the catch-all.

jrouly commented:
Also just an aside, but it probably makes more sense to implement the catch-all as `suggestedBump.getOrElse(catchAll)` rather than in the regex, simply to make it more obvious.

That would even allow downstream users to set their own `catchAll`, in the case that you want `default=minor` and I want `default=patch`, for example. And potentially even `default=error`.

jrouly commented:
Went ahead and implemented my previous comment.

Adds `defaultBump = settingKey[Option[Bump]]` which is used in the case that no commits match any regex. If it's set to `None`, the existing error is thrown. Otherwise, the default bump is returned.

If the default bump is used, a warning log is emitted.

dkichler commented:
I also was glad to stumble across this plugin recently (thanks @pdalpra) before starting efforts of my own toward some of it's features.  I also discovered the limitations this PR addresses and was able to step around them by overriding the regexes to align with conventional commits (and also remove the `.*` catch-all).  Will look forward to the integration and release of these changes (thanks @jrouly) 👍 👍

pdalpra commented:
Thanks for your kind words, even if it didn't fulfill its original use case (built for a previous gig need, but open sourceable), I'm glad it has a second life 🙂
I'll have a look at your PRs this weekend!

pdalpra commented on src/main/scala/autoversion/AutoVersionPlugin.scala:
```diff
@@ -49,12 +49,23 @@ object AutoVersionPlugin extends AutoPlugin {
+
+    if (suggestedBumps.nonEmpty) suggestedBumps.max
+    else default match {
+      case None => sys.error("No commit matches either patterns for bugfix, minor or major bumps !")
```
small nitpicking, but I tend to find throwing the exception directly simpler than using `sys.error`, since there's no indirection and `sys.error` is not that often used in Scala codebases that I know of.

jrouly replied:
Sure, I can switch to `throw`. I like `sys.error` for sbt plugins because it's more terse, but `throw` works great too.

pdalpra commented on src/sbt-test/sbt-autoversion/simple/test:
```diff
@@ -0,0 +1,104 @@
+# Generate project files.
+> reload
```
I don't think this one is required, since you have no "state" you might need to refresh a that point

jrouly replied:
You're right, I always start with `> reload` out of habit but it isn't necessary, can trim.

pdalpra commented on src/sbt-test/sbt-autoversion/simple/test:
```diff
@@ -0,0 +1,104 @@
+$ exec git commit -m 'initial commit'
+$ exec git tag v0.0.0
+
+# Prove sbt can start.
+> reload
+> compile
+> show version
```
Is that needed, given that the first lines of your test run a first command and that the next commands would fail if there was an issue with sbt ?

jrouly replied:
`compile` (or really any initial command) was needed when I was figuring out the `.gitignore` for the test scenario. `show version` was only useful for debugging the tests.

Neither are needed now, I'll delete this whole first chunk.

pdalpra commented on src/test/scala/autoversion/AutoVersionSpec.scala:
```diff
@@ -0,0 +1,61 @@
+import sbtrelease.Version.Bump
+import sbtrelease.{Version, versionFormatError}
+
+class AutoVersionSpec extends FreeSpec with Matchers {
```
Would it be possible to switch to FlatSpec instead, to avoid mixing testing styles ?

jrouly replied:
Sure thing!

pdalpra reviewed (commented):
LGTM, beyond minor nitpickings 👍 
Can you do a run of scalafmt (`scalafmt`/`test:scalafmt`/`scalafmtSbt`)? seems that a bit of formatting got lost

jrouly commented:
I didn't actually realize that scalafmt was in use in this repo. Since the sbt plugin is already installed, I went ahead and added it to your travis CI build as well, hope that's alright.

jrouly commented on src/test/scala/autoversion/model/BumpOrderingSpec.scala:
```diff
@@ -15,25 +15,22 @@ class BumpOrderingSpec extends FlatSpec with Matchers with ScalaCheckDrivenPrope
   it should "always prioritize a major bump" in {
     forAll(bumpListGen) { bumps =>
-      if (bumps.contains(Bump.Major)) {
+      if (bumps.contains(Bump.Major))
```
scalafmt made me do it

jrouly commented:
Pre-weekend bump @pdalpra if there's anything else you'd like to see in this PR :)

pdalpra commented:
Sorry, this week was a busy one 😅
Looking at it now

jrouly commented:
No problem! I appreciate your reviews and openness to PRs :)

pdalpra commented:
LGTM, merging it :)
I’ll try to publish this WE (pending difficulties with publishing plugins to other repositories, since Bintray is shutting down)

jrouly commented:
Where do you expect this plugin (and others, I guess) will end up publishing?

I wrote https://github.com/jrouly/sbt-artifactory for myself since I was also using Bintray, but I quickly migrated off JFrog cloud (jfrog.io) to Sonatype. The plugin still works, though.

I could see `sbt.jfrog.io` being a nice "default" public JFrog repo, maybe? Dunno.

pdalpra commented:
I’ll have a look to see if there is procedure regarding this, maybe a shared access to a common groupId...
I otherwise own a personal domain name, I may end up republish under `me.pdalpra` 🤷

pdalpra commented:
Very sorry for the delay, another busy week 😅 
I'll definitely look into it this week end !

pdalpra commented:
I tried to see if I could get clever and publish it a last time to Bintray, given that the current state of the repository will supposedly be synced to their new repo on Artifactory, but they saw through my cunning plan :sweat_smile: 
I asked Sonatype to grant me access to `me.pdalpra`, I'll publish under this new groupid as soon as it become available, tonight or tomorrow.

jrouly commented:
Haha! That sounds like a plan. I'll keep an eye out for it.

Have you had a chance to think about how you feel about https://github.com/sbt/sbt-autoversion/pull/91, by the way? I'm happy to pull that into a fork / new repo if you don't think it belongs here.

pdalpra commented:
Well, unfortunately plugin maintainers on the sbt organization can't set secrets and use Github Actions + sbt-ci-release, and redoing the whole setup manually is painful...
I'll keep you posted.
I didn't look at your other PR yet unfortunately.

pdalpra commented:
I just release a 2.0.0 version of the plugin, sorry for the wait !
Given it's a first release under a new groupid, it may take a bit longer to see it make its way into Central, but it's here : https://s01.oss.sonatype.org/content/repositories/releases/me/pdalpra/

jrouly commented:
No worries! That's awesome, thanks so much for working through the release process.
